convert2rhel/subscription.py

```diff
@@ -277,28 +376,125 @@ def args(self):
 
         return args
 
-    def __call__(self):
+    @property
+    def connection_opts(self):
         """
-        Run the subscription-manager command.
+        This property is a dbus.Dictionary that contains the connection options for RHSM
+        dbus calls.
 
-        Wrapper for running the subscription-manager command that keeps
-        secrets secure.
+        Set :attr:`server_url` to affect this value.
         """
-        if self.password:
-            loggerinst.debug(
-                "Calling command '%s %s'" % (self.cmd, " ".join(hide_secrets(self.args)))
-            )  # lgtm[py/clear-text-logging-sensitive-data]
-            output, ret_code = utils.run_cmd_in_pty(
-                [self.cmd] + self.args, expect_script=(("[Pp]assword: ", self.password + "\n"),), print_cmd=False
-            )
-        else:
-            # Warning: Currently activation_key can only be specified on the CLI. This is insecure
-            # but there's nothing we can do about it for now. Once subscription-manager issue:
-            # https://issues.redhat.com/browse/ENT-4724 is implemented, we can change both password
-            # and activation_key to use a file-based approach to passing the secrets.
-            output, ret_code = utils.run_subprocess([self.cmd] + self.args, print_cmd=False)
+        connection_opts = {}
+
+        if self.rhsm_hostname:
+            connection_opts["host"] = self.rhsm_hostname
+
+        if self.rhsm_port is not None:
+            connection_opts["port"] = self.rhsm_port
+
+        if self.rhsm_prefix:
+            connection_opts["handler"] = self.rhsm_prefix
+
+        connection_opts = dbus.Dictionary(connection_opts, signature="ss")
+        return connection_opts
+
+    def __call__(self):
+        """
+        Use dbus to register the system with subscription-manager.
+
+        Status of dbus on various platforms:
+            * RHEL6:
+                * DBUS-1.2.24 is present but may not be installed.
+                * Install the dbus package and run /etc/rc.d/init.d/messagebus start
+                * dbus-python-0.83.0 is available
+            * RHEL7:
+                * dbus-1.10.24 is installed and run by default
+                * dbus-python-1.1.9 is available
+            * RHEL8 & RHEL9
+                * dbus-1.12.x is installed and run by default
+                * python3-dbus-1.2.x is available
+
+        .. seealso::
+            Documentation for the subscription-manager dbus API:
+            https://www.candlepinproject.org/docs/subscription-manager/dbus_objects.html
+        """
+        # Note: dbus doesn't understand empty python dicts. Use dbus.Dictionary({}, signature="ss")
+        # if we need one in the future.
+        REGISTER_OPTS_DICT = dbus.Dictionary({"force": True}, signature="ss")
```

Comment:
The signature is incorrect. The registration fails on OL 8.6 with:
```
ERROR:dbus.connection:Unable to set arguments ('', 'username', 'password', dbus.Dictionary({'force': True}, signature=dbus.Signature('ss')), dbus.Dictionary({'host': 'subscription.rhsm.redhat.com'}, signature=dbus.Signature('ss')), 'C') according to signature None: <class 'TypeError'>: Expected a string or unicode object
```

Reply:
This should be fixed.  Correct signature is `dbus.Dictionary({}, signature="sv", variant_level=1)`.  If that doesn't work (It's possible I don't understand variant_level) then we can removethe register_opts dict for this PR and revisit adding it back in later.